src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -465,6 +466,13 @@ public void Write(ArraySegment<byte> data)
         {
             ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
 
+            if (!_canHaveBody)
+            {
+                // Don't write to body for HEAD requests or 101, 204, 205, 304 responses.
+                LogNonBodyRequestWrite(data.Count);
```

Comment:
I think this should throw if there is any actual data. Flush seems fine (though unnecessary), but not anything else.